src/librustc_mir_build/hair/pattern/const_to_pat.rs

```diff
@@ -106,6 +106,12 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
                     traits::NonStructuralMatchTy::Param => {
                         bug!("use of constant whose type is a parameter inside a pattern")
                     }
+                    traits::NonStructuralMatchTy::Dynamic => {
+                        // span_fatal avoids ICE in hair.
+                        self.tcx()
+                            .sess
+                            .span_fatal(self.span, "trait object cannot be used in patterns");
+                    }
```

Comment:
What you want is changes [like these](https://github.com/rust-lang/rust/pull/70872/files#diff-0f60550338d653b11927293df0fef39aL103-R121) (but `Dynamic` instead of `RawPtr`/`FnPtr`).

We should not add any new fatal errors to `rustc`.

Reply:
This fatal error masks an ICE in `fold_const_value_deref`. https://github.com/rust-lang/rust/blob/e82734e56b2a50d38e0937d08f559d15dbe8e46b/src/librustc_mir_build/hair/pattern/_match.rs#L318

Looking into fixing that one rn

Reply:
simple returning val in that case causes an ICE in codegen_fulfill_obligations, fixing this cleanly is currently below my minimum effort/reward ratio
https://github.com/rust-lang/rust/blob/e82734e56b2a50d38e0937d08f559d15dbe8e46b/src/librustc_trait_selection/traits/codegen/mod.rs#L61

Reply:
Can you push your changes? I have a hard time understand how it gets that far.

Worse case you can make `to_pat` return `PatKind::Wild` in that case, but it should be hitting the `span_fatal` lower down.

Reply:
pushed my current progress for now.